src/number_theory/legendre_symbol/add_character.lean

```diff
@@ -63,8 +63,15 @@ The trivial additive character (sending everything to `1`) is `(1 : add_char R R
 @[derive [comm_monoid, inhabited]]
 def add_char : Type (max u v) := (multiplicative R) →* R'
 
-instance add_char.has_coe_to_fun : has_coe_to_fun (add_char R R') (λ x, multiplicative R → R') :=
-monoid_hom.has_coe_to_fun
+open multiplicative
+
+/-- Define coercion to a function so that it includes the move from `R` to `multiplicative R`.
+After we have proved the API lemmas below, we don't need to worry about writing `of_add a`
+when we want to apply an additive character. -/
+instance add_char.has_coe_to_fun : has_coe_to_fun (add_char R R') (λ x, R → R') :=
+{ coe := λ ψ x, ψ.to_fun (of_add x) }
+
+lemma coe_to_fun_apply (ψ : add_char R R') (a : R) : ψ a = ψ.to_fun (of_add a) := rfl
```

Comment:
Could you please add a definition
```
def add_char.to_monoid_hom : (add_char R R') → (multiplicative R →* R') := id
```
and use that on the RHS.
The benefit is that `ψ.to_monoid_hom` will be an actual monoid hom, and so all simp lemmas apply to it, etc.
On the other hand `ψ.to_fun` is not in normal form, and so it will be annoying to work with.
The only simp normal form currently available would be `ψ` itself, but `ψ x` is now ambiguous.

This points to a lingering problem: the definition of `add_char` is not irreducible. You do this because you want to be able to quickly access API for monoid homs when you need it. And I understand that. But it also means that `ψ` can be interpreted in two ways. I think setting up an extra helper function as suggested above will go a long way to make sure that the elaborator doesn't get confused about which `coe_to_fun` should be picked.

Reply:
I have changed that now; I hope I interpreted what you wrote correctly.

Reply:
Yes, looks good. Thanks